### PR TITLE
api/v1: add metrics for each API call made to the cilium-agent

### PR DESCRIPTION
**Pull request by aanm**
The metrics added are in the form of an histogram vec which are scoped
by API path and returned status code.

Signed-off-by: André Martins <andre@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/7112)

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: pkg/metrics/metrics.go [hunk 1]
```diff
@@ -131,6 +131,22 @@ var (
...
+	// LabelAPIReturnCode is the HTTP code returned for that API path
+	LabelAPIReturnCode = "return_code"
+
+	// API interactions
+
+	// APIInteractions is the total time taken to process an API call made
+	// to the cilium-agent
+	APIInteractions = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: Namespace,
+		Name:      "api_process_time",
```
Comment (joestringer) [inline pkg/metrics/metrics.go]: We name other histogram metrics like *_seconds, please do so here too.
```suggestion
		Name:      "api_process_time_seconds",
```

**Review by joestringer — CHANGES_REQUESTED** [review]

File: pkg/metrics/metrics.go [hunk 2]
```diff
@@ -131,6 +131,22 @@ var (
...
+
+	// LabelAPIReturnCode is the HTTP code returned for that API path
+	LabelAPIReturnCode = "return_code"
+
+	// API interactions
+
+	// APIInteractions is the total time taken to process an API call made
+	// to the cilium-agent
+	APIInteractions = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: Namespace,
```
Comment (tgraf) [inline pkg/metrics/metrics.go:151]: Please use the "agent" Subsystem

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -131,6 +131,22 @@ var (
...
+		Name:      "api_process_time",
+		Help:      "API processing time stats labeled by API PATH",
+	}, []string{LabelPath, LabelAPIReturnCode})
```
Comment (tgraf) [inline pkg/metrics/metrics.go]: Also needs method

File: pkg/metrics/middleware.go [hunk 4]
```diff
@@ -24,13 +26,30 @@ import (
...
+	rw.code = code
+	rw.ResponseWriter.WriteHeader(code)
 }
 
 // ServeHTTP implements the http.Handler interface. It records the timestamp
 // this API call began at, then chains to the next handler.
 func (m *APIEventTSHelper) ServeHTTP(r http.ResponseWriter, req *http.Request) {
 	m.TSGauge.SetToCurrentTime()
-	m.Next.ServeHTTP(r, req)
+	t := time.Now()
```
Comment (tgraf) [inline pkg/metrics/middleware.go]: Use `duration := spanstat.Start()` to avoid underflow bug

File: pkg/metrics/middleware.go [hunk 5]
```diff
@@ -24,13 +26,30 @@ import (
...
+	rw := &responderWrapper{ResponseWriter: r}
+	m.Next.ServeHTTP(rw, req)
+	if req != nil && req.URL != nil && req.URL.Path != "" {
+		took := float64(time.Since(t).Nanoseconds()) / float64(time.Second)
```
Comment (tgraf) [inline pkg/metrics/middleware.go]: Use `duration.End(true).Seconds()`

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by aanm** [comment]
test-me-please

File: pkg/metrics/metrics.go [hunk 6]
```diff
@@ -131,6 +134,26 @@ var (
...
+	LabelMethod = "api_method"
+
+	// LabelAPIReturnCode is the HTTP code returned for that API path
+	LabelAPIReturnCode = "return_code"
+
+	// API interactions
+
+	// APIInteractions is the total time taken to process an API call made
+	// to the cilium-agent
+	APIInteractions = prometheus.NewHistogramVec(prometheus.HistogramOpts{
```
Comment (eloycoto) [inline pkg/metrics/metrics.go:150]: Need to add this metric on Doc/config/metrics.rst

**Review by eloycoto — CHANGES_REQUESTED** [review]

**Comment by aanm** [comment]
test-me-please

File: Documentation/configuration/metrics.rst [hunk 7]
```diff
@@ -28,6 +28,11 @@ All metrics are exported under the ``cilium`` Prometheus namespace. When
 running and collecting in Kubernetes they will be tagged with a pod name and
 namespace.
 
+Agent
```
Comment (eloycoto) [inline Documentation/configuration/metrics.rst]: We already have an Agent section.

File: pkg/metrics/middleware.go [hunk 8]
```diff
@@ -24,13 +28,30 @@ import (
...
 // ServeHTTP implements the http.Handler interface. It records the timestamp
 // this API call began at, then chains to the next handler.
 func (m *APIEventTSHelper) ServeHTTP(r http.ResponseWriter, req *http.Request) {
 	m.TSGauge.SetToCurrentTime()
-	m.Next.ServeHTTP(r, req)
+	duration := spanstat.Start()
+	rw := &responderWrapper{ResponseWriter: r}
+	m.Next.ServeHTTP(rw, req)
+	if req != nil && req.URL != nil && req.URL.Path != "" {
+		took := float64(duration.End(true).Total().Nanoseconds()) / float64(time.Second)
```
Comment (eloycoto) [inline pkg/metrics/middleware.go]: Why not? 
`float64(duration.End(true).Total().Seconds())`

File: pkg/metrics/middleware.go [hunk 9]
```diff
@@ -24,13 +28,30 @@ import (
...
+		m.Histogram.WithLabelValues(req.URL.Path, req.Method, strconv.Itoa(rw.code)).Observe(took)
```
Comment (eloycoto) [inline pkg/metrics/middleware.go]: I'm a bit against of this Path in the labels, the Prometheus labels will grow  a  lot and users can have issues with multiple labels. I'm not sure if this is the correct behaviour, let's say that for the getting pod a pod label will be include no? 

```
https://192.168.36.11:6443/api/v1/namespaces/kube-system/pods/cilium-25g7k
```

Also, I would not use response code and use success=true|false

**Review by eloycoto — CHANGES_REQUESTED** [review]

File: pkg/metrics/middleware.go (hunk 9 above)
Comment (tgraf) [reply pkg/metrics/middleware.go]: Agreed. We need to shorten the path to the unique function portion.

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: pkg/metrics/middleware.go [hunk 10]
```diff
@@ -24,13 +28,48 @@ import (
...
 func (m *APIEventTSHelper) ServeHTTP(r http.ResponseWriter, req *http.Request) {
 	m.TSGauge.SetToCurrentTime()
-	m.Next.ServeHTTP(r, req)
+	duration := spanstat.Start()
+	rw := &responderWrapper{ResponseWriter: r}
+	m.Next.ServeHTTP(rw, req)
+	if req != nil && req.URL != nil && req.URL.Path != "" {
+		path := getShortPath(req.URL.Path)
+		took := float64(duration.End(true).Total().Seconds())
+		m.Histogram.WithLabelValues(path, req.Method, strconv.Itoa(rw.code)).Observe(took)
```
Comment (eloycoto) [inline pkg/metrics/middleware.go:73]: I still think that the `response.code` is not the best variable to add here, for me will be true/false.

**Review by eloycoto — CHANGES_REQUESTED** [review]

File: pkg/metrics/middleware.go (hunk 10 above)
Comment (tgraf) [reply pkg/metrics/middleware.go:73]: @eloycoto Are you worried about too many labels? We can use `label_replace()` to turn it into a 4xx and 5xx label in the query function. I think it is valuable to know what type of error is being returned. We have well-documented error codes and a limited number thereof.

File: pkg/metrics/metrics.go [hunk 11]
```diff
@@ -131,6 +134,26 @@ var (
 	// LabelKind is the kind a label
 	LabelKind = "kind"
 
+	// LabelPath is the label for the API path
+	LabelPath = "api_path"
```
Comment (tgraf) [inline pkg/metrics/metrics.go]: the `api_` prefix is not necessary. The label is only valid within the specified metric and the context is clear.

File: pkg/metrics/metrics.go [hunk 12]
```diff
@@ -131,6 +134,26 @@ var (
...
+
+	// LabelMethod is the label for the HTTP method
+	LabelMethod = "api_method"
```
Comment (tgraf) [inline pkg/metrics/metrics.go]: Same here

File: pkg/metrics/metrics.go [hunk 13]
```diff
@@ -131,6 +134,26 @@ var (
...
+
+	// API interactions
+
+	// APIInteractions is the total time taken to process an API call made
+	// to the cilium-agent
+	APIInteractions = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Namespace: Namespace,
+		Subsystem: Agent,
+		Name:      "api_process_time_seconds",
+		Help:      "API processing time stats labeled by API PATH",
```
Comment (tgraf) [inline pkg/metrics/metrics.go]: The description is not entirely clear, let's use `"Duration of processed API calls labeled by path, method and return code."`

**Review by tgraf — CHANGES_REQUESTED** [review]
Starts to look great, please provide the output of `cilium metrics list | grep api` so we can see how it looks like.

File: pkg/metrics/middleware.go (hunk 10 above)
Comment (eloycoto) [reply pkg/metrics/middleware.go:73]: Yep,not too worried about the number of labels, but yes, [0-5]xx works for me

**Comment by aanm** [comment]
> Starts to look great, please provide the output of `cilium metrics list | grep api` so we can see how it looks like.


@tgraf 

```
cilium_agent_api_process_time_seconds                 path="/v1/endpoint" return_code="200" method="DELETE"                        0.876058
cilium_agent_api_process_time_seconds                 path="/v1/ipam" return_code="200" method="DELETE"                            0.000157
cilium_agent_api_process_time_seconds                 path="/v1/config" return_code="200" method="GET"                             0.001809
cilium_agent_api_process_time_seconds                 method="GET" path="/v1/endpoint" return_code="200"                           0.000909
cilium_agent_api_process_time_seconds                 method="GET" path="/v1/healthz" return_code="200"                            0.014038
cilium_agent_api_process_time_seconds                 method="GET" path="/v1/metrics" return_code="200"                            0.036981
cilium_agent_api_process_time_seconds                 method="POST" path="/v1/ipam" return_code="201"                              0.000191
cilium_agent_api_process_time_seconds                 method="PUT" path="/v1/endpoint" return_code="201"                           5.887717
```

**Comment by aanm** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
This will give us fantastic insights

```
cilium_agent_api_process_time_seconds                 method="PUT" path="/v1/endpoint" return_code="201"                           5.887717
```

**Review by nebril — APPROVED** [review]

**Pull request merged**